fastvideo/v1/pipelines/stages/denoising.py

```diff
@@ -0,0 +1,213 @@
+
+        # Get timesteps and calculate warmup steps
+        timesteps = batch.timesteps
+        num_inference_steps = batch.num_inference_steps
+        num_warmup_steps = len(timesteps) - num_inference_steps * self.scheduler.order
+
+        # Create 3D list for mask strategy
+        def dict_to_3d_list(mask_strategy, t_max=50, l_max=60, h_max=24):
+            result = [[[None for _ in range(h_max)] for _ in range(l_max)] for _ in range(t_max)]
+            if mask_strategy is None:
+                return result
+            for key, value in mask_strategy.items():
+                t, l, h = map(int, key.split('_'))
+                result[t][l][h] = value
+            return result
+
+        mask_strategy = dict_to_3d_list(batch.mask_strategy)
+        
+        # Get latents and embeddings
+        latents = batch.latents
+        prompt_embeds = batch.prompt_embeds
+        prompt_embeds_2 = batch.prompt_embeds_2
+        
+        # Run denoising loop
+        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            for i, t in enumerate(timesteps):
+                # Skip if interrupted
+                if hasattr(self, 'interrupt') and self.interrupt:
+                    continue
+
+                # Expand latents for classifier-free guidance
+                latent_model_input = (torch.cat([latents] * 2) if batch.do_classifier_free_guidance else latents)
+                latent_model_input = self.scheduler.scale_model_input(latent_model_input, t)
+
+                # Prepare inputs for transformer
+                t_expand = t.repeat(latent_model_input.shape[0])
+                guidance_expand = (torch.tensor(
+                    [inference_args.embedded_cfg_scale] * latent_model_input.shape[0],
+                    dtype=torch.float32,
+                    device=batch.device,
+                ).to(target_dtype) * 1000.0 if inference_args.embedded_cfg_scale is not None else None)
+                
+                # Predict noise residual
+                with torch.autocast(device_type="cuda", dtype=target_dtype, enabled=autocast_enabled):
+                    # Prepare encoder hidden states
+                    if prompt_embeds_2 is not None and prompt_embeds_2.shape[-1] != prompt_embeds.shape[-1]:
+                        prompt_embeds_2 = torch.nn.functional.pad(
+                            prompt_embeds_2,
+                            (0, prompt_embeds.shape[2] - prompt_embeds_2.shape[1]),
+                            value=0,
+                        ).unsqueeze(1)
+                    encoder_hidden_states = torch.cat([prompt_embeds_2, prompt_embeds], dim=1) if prompt_embeds_2 is not None else prompt_embeds
+                    
+                    # TODO(will): finalize the interface
+                    with set_forward_context(num_step=i, inference_args=inference_args):
+                        # Run transformer
+                        noise_pred = self.transformer(
```

Comment:
This pattern is confusing, because it means this stage cannot be used without pipeline. Ideally every stage should be self-contained. For example

```python
def __init__(self, ...):
    self.transformers = ...

def _call_implementation(self, ...):
    batch = self.transformers(...)
```